src/files.ml

```diff
@@ -3,6 +3,211 @@
+    terminated with a graceful error message. *)
+let set_lib_root : string -> unit = fun path ->
+  try
+    let path = Filename.realpath path in
+    if not (Sys.is_directory path) then
+      exit_with "Invalid library root ([%s] is not a directory)." path;
+    match Pervasives.(!lib_root) with
+    | None    -> Pervasives.(lib_root := Some(path))
+    | Some(_) -> exit_with "The library root was already set."
+  with Sys_error(_) | Invalid_argument(_) ->
+    exit_with "Invalid library root (no such file or directory [%s])." path
+
+(** [default_lib_root ()] returns the default library root curently configured
+    for the system. It depends on the current Opam switch (if any), and it may
+    or may not correspond to an existing directory. *)
+let default_lib_root : unit -> string = fun _ ->
+  match run_process "opam var lib" with
+  | Some([l]) -> Filename.concat l "lambdapi/lib_root"
+  | _         -> "/usr/local/lib/lambdapi/lib_root"
+
+(** [lib_root_path ()] returns the canonical library root path. It corresponds
+    to a valid path to a directory. The function fails gracefully if it is not
+    able to find such a directory. Note that prior to installation, the option
+    ["--lib-root"] must be used so that this function does not fail. *)
+let lib_root_path : unit -> string = fun _ ->
+  match Pervasives.(!lib_root) with Some(path) -> path | None ->
+  let path = default_lib_root () in
+  try
+    let path = Filename.realpath path in
+    if not (Sys.is_directory path) then
+      exit_with "Invalid default library root [%s] (not a directory)." path;
+    path
+  with Sys_error(_) | Invalid_argument(_) ->
+    exit_with "Default library root [%s] does not exist." path
+
+(** [lib_mappings] stores the specified mappings of library paths. *)
+let lib_mappings : ModMap.t Pervasives.ref =
+  Pervasives.ref ModMap.empty
+
+(** [init_lib_root ()] registers the currently set library root as part of our
+    module mapping. This function MUST be called before one can consider using
+    [module_to_file] or [module_path]. *)
+let init_lib_root : unit -> unit = fun _ ->
+  let root = lib_root_path () in
+  Pervasives.(lib_mappings := ModMap.set_root root !lib_mappings)
+
+(** [new_lib_mapping s] attempts to parse [s] as a library mapping of the form
+    ["<modpath>:<path>"]. Then, if module path ["<modpath>"] is not yet mapped
+    to a file path, and if ["<path>"] corresponds to a valid directory, then a
+    new mapping is registered. In case of failure the program terminates and a
+    graceful error message is displayed. *)
+let new_lib_mapping : string -> unit = fun s ->
+  let fail s = exit_with ("Ill-formed argument to \"--map\": " ^^ s ^^ ".") in
+  let (module_path, file_path) =
+    match String.split_on_char ':' s with
+    | [mp; dir] -> (String.split_on_char '.' mp, dir)
+    | _         -> fail "bad syntax (use \"--help\")"
+  in
+  let valid_path_member s =
```

Comment:
It would be better to have a common code with the parser to recognize regular idents.

Reply:
I don't think we want the full syntax of module paths here, the subset I am checking for seems reasonable to me. Anyway, we are changing our parser library so I will reconsider doing that when the PaComb stuff has been merged.